llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp

```diff
@@ -1931,6 +1976,9 @@ struct DSEState {
     if (!Store)
       return false;
 
+    if (dominatingConditionImpliesValue(Def))
```

Comment:
I think we should pass the `Store` instruction here to avoid confusion in the callee. That will also remove a dyn_cast you have there.